lib/sources/bulk_assembly.rb

```diff
@@ -0,0 +1,130 @@
+module Sources
+  class BulkAssembly
+    attr_reader :output
+
+    def initialize(params)
+      @sources = params[:sources]
+      @endpoints = params[:endpoints]
+      @applications = params[:applications]
+
+      # separate out the superkey authentications.
+      @superkeys, @authentications = params[:authentications]&.partition { |auth| auth[:authtype] == "superkey" }
+    end
+
+    def process
+      Source.transaction do
+        @output = {}.tap do |output|
+          # Create the base source(s)
+          output[:sources] = create_sources(@sources)
+
+          # Create the superkey authentications (if there are any)
+          output[:authentications] = create_authentications(@superkeys, output)
+
+          # Create the endpoints
+          output[:endpoints] = create_endpoints(@endpoints, output)
+
+          # Create the applications (sends superkey requests via callback)
+          output[:applications] = create_applications(@applications, output)
+
+          # Create the remaining authentications that aren't superkeys.
+          if output[:authentications]
+            output[:authentications].concat(create_authentications(@authentications, output))
+          else
+            output[:authentications] = create_authentications(@authentications, output)
+          end
+        end
+
+        self
+      rescue => e
+        Rails.logger.error("Error bulk processing from payload: Sources: #{@sources}, Endpoints: #{@endpoints}, Applications: #{@applications}, Authentications: #{@authentications}. Error: #{e}")
+
+        raise
+      end
+    end
+
+    def create_sources(sources)
+      sources&.map do |source|
+        # get the source type by ID or by type string
+        srct = SourceType.find_by(:id => source.delete(:source_type_id)) || SourceType.find_by(:name => source.delete(:type))
+
+        raise ActiveRecord::ActiveRecordError, "Source Type not found" if srct.nil?
+
+        Source.create!(source.merge!(:source_type => srct))
+      end
+    end
+
+    def create_endpoints(endpoints, resources)
+      endpoints&.map do |endpoint|
+        src = find_resource(resources, :sources, endpoint.delete(:source_name))
+
+        Endpoint.create!(endpoint.merge!(:source_id => src.id))
+      end
+    end
+
+    def create_applications(applications, resources)
+      applications&.map do |app|
+        src = find_resource(resources, :sources, app.delete(:source_name))
+        # Get the application by id or lookup by type string
+        appt = ApplicationType.find_by(:id => app.delete(:application_type_id)) || get_application_type(app.delete(:type))
```

Comment:
1) the same question like in create_sources
2) it checks against ApplicationType.name so if it's necessary, it should be consistent with the `create_sources` method and should be called `app.delete(:application_type_name)`

Reply:
Renamed to `application_type_name` for the lookup column.